Bug Report: Flaky unit tests: `TestMonitorClearAllData`, `TestDemotePrimaryStalled`, and `TestMonitorWriteBlocked`

### Overview of the Issue

## Symptoms

Two unit tests fail intermittently on CI, more frequently on faster runners
running with `VT_GO_PARALLEL_VALUE=$(nproc)` (i.e., one test binary per core
in parallel). We have recently seen this a lot in our fork after recently moving to much larger and faster GitHub Actions runners. The failure modes:

### `go/vt/vttablet/tabletmanager/semisyncmonitor TestMonitorClearAllData`

```
=== FAIL: go/vt/vttablet/tabletmanager/semisyncmonitor TestMonitorClearAllData (0.10s)
    monitor_test.go:439:
            Error: "set session lock_wait_timeout=5" does not contain "truncate table _vt.semisync_heartbeat"
            Test:  TestMonitorClearAllData
... INF semisyncmonitor/monitor.go:179 SemiSync Monitor: closing
... ERR mysql/conn.go:1603 conn 1: flush() failed: write unix /tmp/fakesqldb*/fakesqldb.sock->@: write: broken pipe
```

The fakesqldb `QueryLog()` only contains the SET (the first half of the
multi-statement); the TRUNCATE half is missing. The "broken pipe" line is
just cleanup noise from `m.Close()`.

### `go/vt/vttablet/tabletserver TestDemotePrimaryStalled`

```
state_manager_test.go:681:
    Error: Should be empty, but was tabletserver uninitialized
    Test:  TestDemotePrimaryStalled
...
panic: test timed out after 10m0s
    running tests:
        TestDemotePrimaryStalled (9m47s)
```

A `require.Empty` fires on the first `<-ch` read because the streamed health
state still carries the initial `tabletserver uninitialized` error. After
that, `t.FailNow` triggers a deadlock in the deferred cleanup, so the test
hangs for the full 10-minute package timeout, taking down the entire
`tabletserver` package run with it.

## Root causes

### 1. `clearAllData` does not drain its multi-statement results

`semisyncmonitor/monitor.go` issues
`"SET SESSION lock_wait_timeout=5;TRUNCATE TABLE _vt.semisync_heartbeat"` and
calls `ExecuteFetchMulti(..., 0, false)` while discarding the `more` return
value. `ExecuteFetchMulti` only reads the FIRST result; the second OK packet
is left unread on the connection.

On a busy CI runner the server-side `handleComQuery` goroutine can be
descheduled between processing query 1 and query 2. After 100 ms of idle,
the buffered-writer flush timer (`go/mysql/conn.go:457`,
`DefaultFlushDelay = 100ms`) fires asynchronously and ships query 1's OK
packet to the client. The client's `ReadQueryResult` returns immediately,
`ExecuteFetchMulti` returns, `clearAllData` returns, and the test reads
`db.QueryLog()` BEFORE `HandleQuery` for the TRUNCATE has run. So
`db.querylog` only contains the SET.

Compare to the write path on line 379, which already does this correctly:
`conn.Conn.ExecuteFetchMultiDrain(...)` reads all results, which forces the
server to fully complete `handleComQuery` before the client returns.

This is also a real production bug: every successful `clearAllData` leaves
a stray result packet on a pooled connection. The next user of that
connection reads the TRUNCATE's response as its own query result.

### 2. `TestDemotePrimaryStalled` races against `register()`

In `health_streamer.go`, `register()` pushes
`hs.state.CloneVT()` onto the streaming client's channel as soon as the
client subscribes. The initial state has
`HealthError = "tabletserver uninitialized"` and is only cleared by a
`Broadcast() -> ChangeState()` call.

`TestDemotePrimaryStalled` does:

1. `SetServingType(...)` — schedules `go sm.hcticks.Trigger()` (an
   asynchronous Broadcast) inside `setState`.
2. `sm.hcticks.Stop()` — IMMEDIATELY stops the timer; if the queued
   `Trigger()` hasn't run yet, it sees `running=false` and no-ops, so the
   state stays uninit.
3. `wg.Go(...)` — launches the Stream goroutine.
4. `sm.Broadcast()` — explicit broadcast, intended to be the first state
   the test observes.
5. `<-ch` and `require.Empty(...)` — fails when the first message popped
   is the uninit state pushed by `register()`.

The race: if the `wg.Go` goroutine reaches `register()` BEFORE the test
thread reaches `Broadcast()`, the registered channel gets the uninit state
first. With `VT_GO_PARALLEL_VALUE=$(nproc)` the runtime preempts more
aggressively and `register()` wins more often.

The 10-minute hang on failure is a separate bug: `defer wg.Wait()` was
queued before `defer sm.StopService()`, so on `t.FailNow` they unwind LIFO
and `wg.Wait()` blocks waiting for `Stream()` to return — which only
happens once `StopService()` cancels the streamer context. Deadlock until
the package's 10-minute timeout fires.

### 3. `TestMonitorWriteBlocked` races against the write goroutine

`TestMonitorWriteBlocked` calls `m.write()` once in a goroutine and then
polls for `m.inProgressWriteCount > 0`. There is no actual blocking in the
write path: under fakesqldb, `m.write()` runs
`incrementWriteCount → ExecuteFetchMultiDrain → decrementWriteCount` and
all three complete in sub-millisecond time. On a busy CI runner the
polling goroutine may not be scheduled at all during the write's lifetime,
so `Eventually` never observes the transient `count > 0` state.

The previous loop adding the SET/INSERT queries 15 times was also a
misunderstanding: `db.AddQuery` keys into a `map[string]*ExpectedResult`,
so the loop was just overwriting the same entry repeatedly. There is no
"queue of queries" in fakesqldb.

### Reproduction Steps

Use a large (e.g. 32 vCPU) machine:

```
export VT_GO_PARALLEL_VALUE=32
make unit_test_race
```

### Binary Version

```sh
main
```

### Operating System and Environment details

```sh
N/A
```

### Log Fragments

```sh

```

--- a/go/vt/vttablet/tabletmanager/semisyncmonitor/monitor.go
+++ b/go/vt/vttablet/tabletmanager/semisyncmonitor/monitor.go
@@ -416,7 +416,7 @@ func (m *Monitor) clearAllData() {
 		return
 	}
 	defer conn.Recycle()
-	_, _, err = conn.Conn.ExecuteFetchMulti(m.addLockWaitTimeout(m.bindSideCarDBName(semiSyncHeartbeatClear)), 0, false)
+	err = conn.Conn.ExecuteFetchMultiDrain(m.addLockWaitTimeout(m.bindSideCarDBName(semiSyncHeartbeatClear)))
 	if err != nil {
 		m.errorCount.Add(1)
 		log.Error(fmt.Sprintf("SemiSync Monitor: failed to clear semisync_heartbeat table: %v", err))

--- a/go/vt/vttablet/tabletmanager/semisyncmonitor/monitor_test.go
+++ b/go/vt/vttablet/tabletmanager/semisyncmonitor/monitor_test.go
@@ -669,8 +669,14 @@ func TestMonitorWriteBlocked(t *testing.T) {
 	db.AddQuery("INSERT INTO _vt.semisync_heartbeat (ts) VALUES (NOW())", &sqltypes.Result{})
 	// Block the INSERT so we have a deterministic window in which write() is in
 	// progress; otherwise the goroutine can complete before the assertion below
-	// observes inProgressWriteCount > 0 on a busy CI runner.
+	// observes inProgressWriteCount > 0 on a busy CI runner. release() is
+	// idempotent and registered with t.Cleanup so a require.* failure can't
+	// leave the INSERT goroutine wedged inside BeforeFunc and deadlock
+	// db.Close() during teardown.
 	unblock := make(chan struct{})
+	var unblockOnce sync.Once
+	release := func() { unblockOnce.Do(func() { close(unblock) }) }
+	t.Cleanup(release)
 	db.SetBeforeFunc("INSERT INTO _vt.semisync_heartbeat (ts) VALUES (NOW())", func() {
 		<-unblock
 	})
@@ -691,7 +697,7 @@ func TestMonitorWriteBlocked(t *testing.T) {
 	}, 15*time.Second, time.Millisecond)
 
 	// Let the INSERT proceed and check that the write finished successfully.
-	close(unblock)
+	release()
 	require.Eventually(t, func() bool {
 		return writeFinished.Load()
 	}, 10*time.Second, 100*time.Millisecond)

--- a/go/vt/vttablet/tabletmanager/semisyncmonitor/monitor_test.go
+++ b/go/vt/vttablet/tabletmanager/semisyncmonitor/monitor_test.go
@@ -665,28 +665,33 @@ func TestMonitorWriteBlocked(t *testing.T) {
 	m.mu.Unlock()
 
 	// ExecuteFetchMulti will execute each statement separately, so we need to add SET query and INSERT query.
-	// Add them multiple times so the writes can execute.
-	for range maxWritesPermitted {
-		db.AddQuery("SET SESSION lock_wait_timeout=1", &sqltypes.Result{})
-		db.AddQuery("INSERT INTO _vt.semisync_heartbeat (ts) VALUES (NOW())", &sqltypes.Result{})
-	}
+	db.AddQuery("SET SESSION lock_wait_timeout=1", &sqltypes.Result{})
+	db.AddQuery("INSERT INTO _vt.semisync_heartbeat (ts) VALUES (NOW())", &sqltypes.Result{})
+	// Block the INSERT so we have a deterministic window in which write() is in
+	// progress; otherwise the goroutine can complete before the assertion below
+	// observes inProgressWriteCount > 0 on a busy CI runner.
+	unblock := make(chan struct{})
+	db.SetBeforeFunc("INSERT INTO _vt.semisync_heartbeat (ts) VALUES (NOW())", func() {
+		<-unblock
+	})
 
-	// Do a write, which we expect to block.
+	// Do a write, which we expect to block on the INSERT.
 	var writeFinished atomic.Bool
 	go func() {
 		m.write()
 		writeFinished.Store(true)
 	}()
 
-	// We should see the number of writers increase briefly, before it completes.
+	// We should see the number of writers increase while the INSERT is blocked.
 	require.Zero(t, m.errorCount.Get())
 	require.Eventually(t, func() bool {
 		m.mu.Lock()
 		defer m.mu.Unlock()
 		return m.inProgressWriteCount > 0
-	}, 15*time.Second, 1*time.Microsecond)
+	}, 15*time.Second, time.Millisecond)
 
-	// Check that the writes finished successfully.
+	// Let the INSERT proceed and check that the write finished successfully.
+	close(unblock)
 	require.Eventually(t, func() bool {
 		return writeFinished.Load()
 	}, 10*time.Second, 100*time.Millisecond)

--- a/go/vt/vttablet/tabletserver/state_manager_test.go
+++ b/go/vt/vttablet/tabletserver/state_manager_test.go
@@ -658,7 +658,6 @@ func TestStateManagerNotify(t *testing.T) {
 
 func TestDemotePrimaryStalled(t *testing.T) {
 	sm := newTestStateManager()
-	defer sm.StopService()
 	err := sm.SetServingType(topodatapb.TabletType_PRIMARY, testNow, StateServing, "")
 	require.NoError(t, err)
 	// Stopping the ticker so that we don't get unexpected health streams.
@@ -673,7 +672,16 @@ func TestDemotePrimaryStalled(t *testing.T) {
 		})
 		assert.Contains(t, err.Error(), "tabletserver is shutdown")
 	})
+	// Order matters: StopService cancels the streamer's context, which lets
+	// Stream return so wg.Wait can complete. Defers run LIFO.
 	defer wg.Wait()
+	defer sm.StopService()
+
+	// register() pushes the current state onto the channel as soon as the
+	// streaming client subscribes, and that value is racy with respect to
+	// SetServingType's async Broadcast trigger. Drain it so the assertions
+	// below only inspect states produced by our explicit Broadcast calls.
+	<-ch
 
 	// Send a broadcast message and check we have no error there.
 	sm.Broadcast()
@@ -688,9 +696,6 @@ func TestDemotePrimaryStalled(t *testing.T) {
 	// Verify that we can't start a new request once we have a demote primary stalled.
 	err = sm.StartRequest(context.Background(), &querypb.Target{TabletType: topodatapb.TabletType_PRIMARY}, false)
 	require.ErrorContains(t, err, "operation not allowed in state NOT_SERVING")
-
-	// Stop the state manager.
-	sm.StopService()
 }
 
 func TestRefreshReplHealthLocked(t *testing.T) {

